client/python/gradio_client/client.py

```diff
@@ -1557,7 +1557,8 @@ def __next__(self) -> tuple | Any:
                     o = self.communicator.job.outputs[self._counter]
                     self._counter += 1
                     return o
-                if self.communicator.job.latest_status.code == Status.FINISHED:
+                if self.communicator.job.latest_status.code == Status.FINISHED and \
```

Comment:
Thanks @pseudotensor ! This looks good but it won't fix the following point you raised in the issue right?

> Issue is that by time code reaches the first conditional, the number of job.outputs may already be 2+ while self._counter=0 at first.

I think we can copy over the new entries of `communicator.job.outputs` to a temporary list and then pop from that. Something like this:


```python
        while True:
            with self.communicator.lock:
                if len(self.communicator.job.outputs) > self._counter:
                    for i in range(self._counter, len(self.communicator.job.outputs)):
                        self._outputs.append(self.communicator.job.outputs[i])
                    self._counter = len(self.communicator.job.outputs)
                if self._outputs:
                    return self._outputs.pop(0)
                if self.communicator.job.latest_status.code == Status.FINISHED and not self._outputs:
                    raise StopIteration()
                time.sleep(0.001)
```

Reply:
What I mentioned was that your team already fixed the first issue I mentioned in gradio 4 updates.

Reply:
To be clear, the only issue with the first conditional was a > vs. >=

Reply:
Ah gotcha - ok looks good to me